pinot-plugins/pinot-minion-tasks/pinot-minion-builtin-tasks/src/main/java/org/apache/pinot/plugin/minion/tasks/upsertcompactmerge/UpsertCompactMergeTaskGenerator.java

```diff
@@ -0,0 +1,419 @@
+        long totalInvalidDocs = validDocIdsMetadata.getTotalInvalidDocs();
+        long totalValidDocs = validDocIdsMetadata.getTotalValidDocs();
+
+        // Skip segments if the crc from zk metadata and server does not match. They may be getting reloaded.
+        if (segment.getCrc() != Long.parseLong(validDocIdsMetadata.getSegmentCrc())) {
+          LOGGER.warn("CRC mismatch for segment: {}, (segmentZKMetadata={}, validDocIdsMetadata={})", segmentName,
+              segment.getCrc(), validDocIdsMetadata.getSegmentCrc());
+          continue;
+        }
+
+        // segments eligible for deletion with no valid records
+        long totalDocs = validDocIdsMetadata.getTotalDocs();
+        if (totalInvalidDocs == totalDocs) {
+          segmentsForDeletion.add(segmentName);
+        } else if (alreadyMergedSegments.contains(segmentName)) {
+          LOGGER.debug("Segment {} already merged. Skipping it for {}", segmentName,
+              MinionConstants.UpsertCompactMergeTask.TASK_TYPE);
+          continue;
+        } else {
+          Integer partitionID = SegmentUtils.getPartitionIdFromRealtimeSegmentName(segmentName);
+          if (partitionID == null) {
+            LOGGER.warn("Partition ID not found for segment: {}, skipping it for {}", segmentName,
+                MinionConstants.UpsertCompactMergeTask.TASK_TYPE);
+            continue;
+          }
+          segmentsEligibleForCompactMerge.computeIfAbsent(partitionID, k -> new ArrayList<>())
+              .add(new SegmentMergerMetadata(segment, totalValidDocs, totalInvalidDocs));
+        }
+        break;
+      }
+    }
+
+    segmentsEligibleForCompactMerge.forEach((partitionID, segmentList) -> segmentList.sort((o1, o2) -> {
+      // Sort primarily by creationTime in ascending order
+      if (o1.getSegmentZKMetadata().getCreationTime() < o2.getSegmentZKMetadata().getCreationTime()) {
+        return -1;
+      } else if (o1.getSegmentZKMetadata().getCreationTime() == o2.getSegmentZKMetadata().getCreationTime()) {
+        return 0;
+      }
+      return 1;
+    }));
+
+    // Map to store the result: each key (partition) will have a list of groups
+    Map<Integer, List<List<SegmentMergerMetadata>>> groupedSegments = new HashMap<>();
+
+    // Iterate over each partition and process its segments list
+    for (Map.Entry<Integer, List<SegmentMergerMetadata>> entry : segmentsEligibleForCompactMerge.entrySet()) {
+      int partitionID = entry.getKey();
+      List<SegmentMergerMetadata> segments = entry.getValue();
+      // task config thresholds
+      long validDocsThreshold = Long.parseLong(taskConfigs.getOrDefault(MinionConstants
+              .UpsertCompactMergeTask.MAX_NUM_RECORDS_PER_SEGMENT_KEY,
+          String.valueOf(MinionConstants.UpsertCompactMergeTask.DEFAULT_MAX_NUM_RECORDS_PER_SEGMENT)));
+      long maxRecordsPerTask = Long.parseLong(taskConfigs.getOrDefault(MinionConstants
```

Comment:
Let's leave a TODO for finalizing the semantics. Specifically, we should rely on segment size as our primary guide instead of max rows per task. We could do it in a separate PR which can help drive a discussion too.